src/robot/running/bodyrunner.py

```diff
@@ -654,23 +661,34 @@ def create(cls, limit, on_limit_message, variables):
             if count <= 0:
                 raise DataError(f"Invalid WHILE loop limit: Iteration count must be "
                                 f"a positive integer, got '{count}'.")
-            return IterationCountLimit(count, on_limit_message)
+            return IterationCountLimit(count, on_limit, on_limit_message)
         try:
             secs = timestr_to_secs(value)
         except ValueError as err:
             raise DataError(f'Invalid WHILE loop limit: {err.args[0]}')
         else:
-            return DurationLimit(secs, on_limit_message)
+            return DurationLimit(secs, on_limit, on_limit_message)
+
+    @classmethod
+    def parse_on_limit(cls, variables, on_limit):
+        if on_limit is None:
+            return None
+        on_limit = variables.replace_string(on_limit)
+        if on_limit.lower() not in ['pass', 'fail']:
+            msg = f"Invalid WHILE loop on_limit: must be one of 'pass', 'fail', got '{on_limit}'."
+            raise DataError(msg)
+        return on_limit.lower()
 
     def limit_exceeded(self):
         if self.on_limit_message:
-            raise ExecutionFailed(self.on_limit_message)
+            raise LimitExceeded(self.on_limit, self.on_limit_message)
         else:
-            raise ExecutionFailed(f"WHILE loop was aborted because "
-                                  f"it did not finish "
-                                  f"within the limit of {self}. "
-                                  f"Use the 'limit' argument to "
-                                  f"increase or remove the limit if needed.")
+            raise LimitExceeded(self.on_limit,
+                                f"WHILE loop was aborted because "
+                                f"it did not finish "
+                                f"within the limit of {self}. "
+                                f"Use the 'limit' argument to "
+                                f"increase or remove the limit if needed.")
```

Comment:
It looks like this could be formatted to take a little less vertical space. Just making individual text lines longer (up to 88 chars) ought to be enough, but formatting like
```python
raise LimitExceeded(
    ...
)
```
is an option as well.